File: android/app/src/main/kotlin/com/findus/crowdcompass/engine/ProtocolWire.kt
package com.findus.crowdcompass.engine

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Byte-faithful v3 codec port of core/protocol.py. */
class ProtocolWire {

    class ProtocolError(message: String) : Exception(message)

    fun encode(pkt: RelayPacket): ByteArray {
        var flags = 0
        var authBlock = ByteArray(0)
        if (pkt.authTag != null) {
            flags = flags or 0x04
            authBlock = pkt.authTag.copyOf(8).also { if (it.size < 8) it.fill(0, it.size, 8) }
        }
        val timestamp = packTime(pkt.timestamp)
        val header = ByteBuffer.allocate(35).order(ByteOrder.BIG_ENDIAN)
            .put(Engine.PROTOCOL_VERSION.toByte())
            .put(pkt.messageType.wire.toByte())
            .put(packId(pkt.eventId?.value.orEmpty(), 16))
            .put(packId(pkt.sender?.value.orEmpty(), 8))
            .putShort(sourceVersion(pkt.sourceVersion))
            .put(pkt.hop.coerceIn(0, 255).toByte())
            .put(pkt.ttl.coerceIn(0, 255).toByte())
            .putInt(timestamp)
            .put(flags.toByte())
            .array()
        val out = ByteArrayOutputStream(header.size + authBlock.size)
        out.write(header)
        out.write(authBlock)
        return out.toByteArray()
    }

    fun decode(blob: ByteArray): RelayPacket {
        if (blob.size < 35) throw ProtocolError("header needs 35 bytes, got ${blob.size}")
        val h = ByteBuffer.wrap(blob).order(ByteOrder.BIG_ENDIAN)
        val ver = h.get().toInt() and 0xFF
        if (ver != Engine.PROTOCOL_VERSION) {
            throw ProtocolError("unsupported version $ver (this codec: ${Engine.PROTOCOL_VERSION})")
        }
        val mtype = MessageType.fromWire(h.get().toInt() and 0xFF)
            ?: throw ProtocolError("unknown message type")
        val eventBytes = ByteArray(16).also { h.get(it) }
        val senderBytes = ByteArray(8).also { h.get(it) }
        val srcVer = h.short.toInt() and 0xFFFF
        val hop = h.get().toInt() and 0xFF
        val ttl = h.get().toInt() and 0xFF
        val ts = h.int.toLong() and 0xFFFFFFFFL
        val flags = h.get().toInt() and 0xFF
        var pos = 35
        if (flags and 0x01 != 0) pos = skipMeasurement(blob, pos)
        if (flags and 0x02 != 0) pos = skipMeasurement(blob, pos)
        if (flags and 0x04 != 0) {
            if (pos + Engine.AUTH_TAG_BYTES > blob.size) {
                throw ProtocolError("auth tag truncated")
            }
            val auth = blob.copyOfRange(pos, pos + Engine.AUTH_TAG_BYTES)
            return RelayPacket(
                messageType = mtype,
                eventId = unpad(eventBytes)?.let { SosId(it) },
                sender = unpad(senderBytes)?.let { NodeId(it) },
                sourceVersion = srcVer,
                hop = hop,
                ttl = ttl,
                timestamp = if (ts != 0L) ts + Engine.EPOCH else Engine.TIME_NONE,
                authTag = auth,
            )
        }
        return RelayPacket(
            messageType = mtype,
            eventId = unpad(eventBytes)?.let { SosId(it) },
            sender = unpad(senderBytes)?.let { NodeId(it) },
            sourceVersion = srcVer,
            hop = hop,
            ttl = ttl,
            timestamp = if (ts != 0L) ts + Engine.EPOCH else Engine.TIME_NONE,
        )
    }

    private fun skipMeasurement(blob: ByteArray, pos: Int): Int {
        if (pos + 8 > blob.size) throw ProtocolError("measurement block truncated")
        var p = pos + 8
        val hasDist = blob[p - 2].toInt() != 0
        val hasDir = blob[p - 1].toInt() != 0
        if (hasDist) p += 4
        if (hasDir) p += 4
        if (p > blob.size) throw ProtocolError("measurement trailing data truncated")
        return p
    }

    private fun packId(value: String, width: Int): ByteArray {
        val raw = value.toByteArray(Charsets.UTF_8)
        val out = ByteArray(width)
        for (i in 0 until minOf(raw.size, width)) out[i] = raw[i]
        return out
    }

    private fun unpad(raw: ByteArray): String? {
        var end = raw.size
        while (end > 0 && raw[end - 1] == 0.toByte()) end--
        if (end == 0) return null
        return String(raw, 0, end, Charsets.UTF_8)
    }

    private fun packTime(t: Double): Int {
        if (t == Engine.TIME_NONE || t <= 0.0) return 0
        val v = Math.round(t - Engine.EPOCH)
        return v.coerceIn(0, 0xFFFFFFFFL).toInt()
    }

    private fun sourceVersion(v: Int): Short =
        v.coerceIn(0, 0xFFFF).toShort()
}

File: ios/Sources/FindUsEngine/ProtocolWire.swift
import Foundation

/// Byte-faithful v3 codec (port of core/protocol.py).
public struct ProtocolWire {

    public struct ProtocolError: Error, CustomStringConvertible {
        public let message: String
        public var description: String { message }
    }

    public init() {}

    public func encode(_ pkt: RelayPacket) -> Data {
        var flags: UInt8 = 0
        var authBlock = Data()
        if let tag = pkt.authTag {
            flags |= 0x04
            authBlock = tag.prefix(8)
            if authBlock.count < EngineConstants.authTagBytes {
                authBlock.append(Data(repeating: 0, count: EngineConstants.authTagBytes - authBlock.count))
            }
        }
        var header = Data()
        header.append(EngineConstants.protocolVersion)
        header.append(UInt8(pkt.messageType.rawValue))
        header.append(packID(pkt.eventId?.value ?? "", width: 16))
        header.append(packID(pkt.sender?.value ?? "", width: 8))
        header.append(packU16(pkt.sourceVersion.clamped(to: 0, 0xFFFF)))
        header.append(UInt8(pkt.hop.clamped(to: 0, 0xFF)))
        header.append(UInt8(pkt.ttl.clamped(to: 0, 0xFF)))
        header.append(packU32(packTime(pkt.timestamp)))
        header.append(flags)
        header.append(authBlock)
        return header
    }

    public func decode(_ blob: Data) throws -> RelayPacket {
        guard blob.count >= 35 else {
            throw ProtocolError(message: "header needs 35 bytes, got \(blob.count)")
        }
        var pos = 0
        let ver = blob[pos]; pos += 1
        guard ver == EngineConstants.protocolVersion else {
            throw ProtocolError(message: "unsupported version \(ver)")
        }
        let mtype = blob[pos]; pos += 1
        guard let mt = MessageType(wire: mtype) else {
            throw ProtocolError(message: "unknown message type \(mtype)")
        }
        let eventBytes = blob[pos..<(pos + 16)]; pos += 16
        let senderBytes = blob[pos..<(pos + 8)]; pos += 8
        let srcVer = unpackU16(blob[pos..<(pos + 2)]); pos += 2
        let hop = Int(blob[pos]); pos += 1
        let ttl = Int(blob[pos]); pos += 1
        let ts = unpackU32(blob[pos..<(pos + 4)]); pos += 4
        let flags = blob[pos]; pos += 1
        if flags & 0x01 != 0 { pos = try skipMeasurement(blob, at: pos) }
        if flags & 0x02 != 0 { pos = try skipMeasurement(blob, at: pos) }
        var auth: Data? = nil
        if flags & 0x04 != 0 {
            guard pos + EngineConstants.authTagBytes <= blob.count else {
                throw ProtocolError(message: "auth tag truncated")
            }
            auth = blob[pos..<(pos + EngineConstants.authTagBytes)]
        }
        return RelayPacket(
            messageType: mt,
            eventId: unpad(eventBytes).map(SosId.init),
            sender: unpad(senderBytes).map(NodeId.init),
            sourceVersion: Int(srcVer),
            hop: hop,
            ttl: ttl,
            timestamp: ts != 0 ? Double(ts) + EngineConstants.epoch : EngineConstants.timeNone,
            authTag: auth)
    }

    private func skipMeasurement(_ blob: Data, at pos0: Int) throws -> Int {
        guard pos0 + 8 <= blob.count else {
            throw ProtocolError(message: "measurement block truncated")
        }
        var pos = pos0 + 8
        let hasDist = blob[pos - 2] != 0
        let hasDir = blob[pos - 1] != 0
        if hasDist { pos += 4 }
        if hasDir { pos += 4 }
        guard pos <= blob.count else {
            throw ProtocolError(message: "measurement trailing data truncated")
        }
        return pos
    }

    private func packID(_ value: String, width: Int) -> Data {
        var out = Data(repeating: 0, count: width)
        let raw = Data(value.utf8)
        for i in 0..<min(raw.count, width) { out[i] = raw[raw.index(raw.startIndex, offsetBy: i)] }
        return out
    }

    private func unpad(_ raw: Data) -> String? {
        var end = raw.count
        while end > 0 && raw[raw.index(raw.startIndex, offsetBy: end - 1)] == 0 { end -= 1 }
        guard end > 0 else { return nil }
        return String(data: raw.subdata(in: raw.startIndex..<raw.index(raw.startIndex, offsetBy: end)), encoding: .utf8)
    }

    private func packTime(_ t: Double) -> UInt32 {
        guard t != EngineConstants.timeNone, t > 0 else { return 0 }
        let v = (t - EngineConstants.epoch).rounded()
        return UInt32(clamping: Int64(v))
    }

    private func packU16(_ v: Int) -> Data {
        Data([UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)])
    }
    private func packU32(_ v: UInt32) -> Data {
        Data([UInt8((v >> 24) & 0xFF), UInt8((v >> 16) & 0xFF), UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)])
    }
    private func unpackU16(_ d: Data) -> UInt16 {
        UInt16(d[d.startIndex]) << 8 | UInt16(d[d.index(d.startIndex, offsetBy: 1)])
    }
    private func unpackU32(_ d: Data) -> UInt32 {
        let v = d.map { UInt32($0) }
        return v[0] << 24 | v[1] << 16 | v[2] << 8 | v[3]
    }
}

private extension FixedWidthInteger {
    func clamped(to range: ClosedRange<Self>) -> Self {
        min(max(self, range.lowerBound), range.upperBound)
    }
}